File: sorting/bubble_sort/bubble.js
var unsorted_array = [5, 3, 42, 398, 28, 2, 20, 5, 2, 0, -20, 4, 2, 4],
    result;

function swap(items, firstIndex, secondIndex) {
  var temp = items[firstIndex];
  items[firstIndex] = items[secondIndex];
  items[secondIndex] = temp;
};

function bubbleSort(array) {
  var i = 0,
      stop, j, len = array.length;

  for(i; i < len; i++) {
    for(j=0, stop=len-i; j < stop; j++) {
      if(array[j] > array[j + 1]){
        swap(array, j, j+1);
        console.log(array);
      }
    }
  }
  return array
};

result = bubbleSort(unsorted_array);
console.log(result);

File: sorting/bubble_sort/lib/bubble.rb
def bubble_sort(list)
  return list if list.size <= 1 # already sorted

  loop do
    swapped = false
    0.upto(list.size-2) do |i|
      if list[i] > list[i+1]
        list[i], list[i+1] = list[i+1], list[i] # swap values
        swapped = true
      end
    end
    break unless swapped
  end

  list
end

# unsorted_array = [5, 3, 42, 398, 28, 2, 20, 5, 2, 0, -20, 4]

# puts bubble_sort(unsorted_array)
